src/commands/farm.rs

```diff
@@ -101,26 +105,41 @@ pub(crate) async fn farm(is_verbose: bool, executor: bool) -> Result<()> {
 
 #[instrument]
 async fn wait_on_farmer(
-    maybe_handles: Option<(JoinHandle<()>, JoinHandle<()>)>,
+    mut maybe_handles: MaybeHandles,
     farmer: Arc<Farmer>,
-    node: Node,
+    node: Arc<Node>,
 ) -> Result<()> {
     // node subscription can be gracefully closed with `ctrl_c` without any problem
     // (no code needed). We need graceful closing for farmer subscriptions.
-    signal::ctrl_c().await?;
-    println!(
-        "\nWill try to gracefully exit the application now. If you press ctrl+c again, it will \
-         try to forcefully close the app!"
-    );
-
-    // closing the subscriptions if there are any
-    if let Some((plotting_handle, solution_handle)) = maybe_handles.as_ref() {
-        plotting_handle.abort();
-        solution_handle.abort();
+    if let Some((_plotting_handle, solution_handle)) = maybe_handles.as_mut() {
+        tokio::select! {
+            _ = signal::ctrl_c() => {
+               println!(
+                    "\nWill try to gracefully exit the application now. Please wait for a couple of seconds... If you press ctrl+c again, it will \
+                    try to forcefully close the app!"
+                );
+                let (plotting_handle, solution_handle) = maybe_handles.as_ref().expect("check is up there");
+                plotting_handle.abort();
+                solution_handle.abort();
+            }
+            res = solution_handle => {
+                // first level is join handle, second layer is actual result from the task
+                match res {
+                    Ok(Ok(())) => unreachable!("solution subscription never ends"),
+                    Ok(Err(err)) => return Err(eyre!("solution subscription crashed with err: {err}")),
+                    Err(err) => return Err(eyre!("couldn't join subscription handle with err: {err:?}"))
+                }
+            }
+            // cannot inspect plotting subscription for errors, since it may end and quit from
+            // `tokio::select`
+        }
+    } else {
```

Comment:
This one works:
```suggestion
        if let Some((plotting_handle, solution_handle)) = maybe_handles.as_mut() {
            futures::select! {
            _ = signal::ctrl_c().fuse() => {
               println!(
                    "\nWill try to gracefully exit the application now. Please wait for a couple of seconds... If you press ctrl+c again, it will \
                    try to forcefully close the app!"
                );
                plotting_handle.abort();
                solution_handle.abort();
            }
            res = solution_handle.fuse() => {
                return res.context("couldn't join subscription handle")?.context("solution subscription crashed");
            }
        }
    } else {
```